File: easy/reverse-integer/cpp/reverse_integer.cpp
#include <limits>

// Reverse Integer - C++17
// Time: O(k), Space: O(1)

int reverseInteger(int x) {
    const int INT_MIN = std::numeric_limits<int>::min();
    const int INT_MAX = std::numeric_limits<int>::max();

    long long a = x; // use 64-bit to safely take abs and accumulate
    int sign = 1;
    if (a < 0) { sign = -1; a = -a; }

    long long rev = 0;
    while (a != 0) {
        int digit = static_cast<int>(a % 10);
        a /= 10;
        if (rev > (static_cast<long long>(INT_MAX) - digit) / 10) return 0;
        rev = rev * 10 + digit;
    }
    rev *= sign;
    if (rev < INT_MIN || rev > INT_MAX) return 0;
    return static_cast<int>(rev);
}

File: easy/reverse-integer/typescript/reverseInteger.ts
// Reverse Integer - TypeScript implementation
// Time: O(k), Space: O(1)

export function reverseInteger(x: number): number {
  const INT_MIN = -2147483648;
  const INT_MAX = 2147483647;

  const sign = x < 0 ? -1 : 1;
  x = Math.abs(Math.trunc(x));
  let rev = 0;

  while (x !== 0) {
    const digit = x % 10;
    x = Math.trunc(x / 10);
    if (rev > Math.trunc((INT_MAX - digit) / 10)) return 0;
    rev = rev * 10 + digit;
  }

  rev *= sign;
  if (rev < INT_MIN || rev > INT_MAX) return 0;
  return rev;
}
